pkg/kubelet/cm/memorymanager/policy_single_numa.go

```diff
@@ -17,50 +17,340 @@ limitations under the License.
 package memorymanager
 
 import (
+	"fmt"
+
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog"
+	corehelper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	v1qos "k8s.io/kubernetes/pkg/apis/core/v1/helper/qos"
 	"k8s.io/kubernetes/pkg/kubelet/cm/memorymanager/state"
 	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/bitmask"
 )
 
 const policyTypeSingleNUMA policyType = "single-numa"
 
 // TODO: need to implement all methods
 
+type reservedMemory map[int]map[v1.ResourceName]uint64
+
 // SingleNUMAPolicy is implementation of the policy interface for the single NUMA policy
 // TODO: need to re-evaluate what field we really need
 type singleNUMAPolicy struct {
+	// machineInfo contains machine memory related information
+	machineInfo *cadvisorapi.MachineInfo
+	// reserved contains memory that reserved for kube
+	systemReserved reservedMemory
 	// topology manager reference to get container Topology affinity
 	affinity topologymanager.Store
 }
 
 var _ Policy = &singleNUMAPolicy{}
 
 // NewPolicySingleNUMA returns new single NUMA policy instance
-func NewPolicySingleNUMA() Policy {
-	return &singleNUMAPolicy{}
+func NewPolicySingleNUMA(machineInfo *cadvisorapi.MachineInfo, reserved reservedMemory, affinity topologymanager.Store) Policy {
+	// TODO: check if we have enough reserved memory for the system
+	//for _, node := range reserved {
+	//	if node[v1.ResourceMemory] == 0 {
+	//
+	//	}
+	//}
+	return &singleNUMAPolicy{
+		machineInfo:    machineInfo,
+		systemReserved: reserved,
+		affinity:       affinity,
+	}
 }
 
 func (p *singleNUMAPolicy) Name() string {
 	return string(policyTypeSingleNUMA)
 }
 
 func (p *singleNUMAPolicy) Start(s state.State) error {
+	if err := p.validateState(s); err != nil {
+		klog.Errorf("[memorymanager] Invalid state: %v, please drain node and remove policy state file", err)
+		return err
+	}
 	return nil
 }
 
 // Allocate call is idempotent
 func (p *singleNUMAPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Container) error {
+	// allocate the memory only for guaranteed pods
+	if v1qos.GetPodQOS(pod) != v1.PodQOSGuaranteed {
+		return nil
+	}
+
+	klog.Infof("[memorymanager] Allocate (pod: %s, container: %s)", pod.Name, container.Name)
+	if blocks := s.GetMemoryBlocks(string(pod.UID), container.Name); blocks != nil {
+		klog.Infof("[memorymanager] Container already present in state, skipping (pod: %s, container: %s)", pod.Name, container.Name)
+		return nil
+	}
+
+	// Call Topology Manager to get the aligned affinity across all hint providers.
+	hint := p.affinity.GetAffinity(string(pod.UID), container.Name)
+	klog.Infof("[memorymanager] Pod %v, Container %v Topology Affinity is: %v", pod.UID, container.Name, hint)
+
+	if !hint.Preferred {
+		return fmt.Errorf("[memorymanager] failed to get preferred NUMA affinity")
```

Comment:
I added all these checks, just in case when you will run topology manager with the other policy than `single-numa-node` and memory manager with `single-numa`, but in general, if topology manager does not have a preferred hint, it will just filter this hint, so if we will have general validation on the kubelet start that will verify alignment of topology and memory managers policy, we do not really need this check